portal.go

```diff
@@ -907,8 +907,81 @@ func (portal *Portal) handleSignalGroupChange(source *User, sender *Puppet, grou
 		Uint32("new_revision", groupMeta.GetRevision()).
 		Logger()
 	ctx := log.WithContext(context.TODO())
-	// TODO handle group changes properly
-	portal.UpdateInfo(ctx, source, nil, groupMeta.GetRevision())
+	groupChange, err := source.Client.GetGroupChange(ctx, groupMeta, types.GroupIdentifier(portal.ChatID))
+	if err != nil {
+		log.Err(err).Msg("Handling GroupChange failed")
+		return
+	}
+	if groupChange.ModifyTitle != nil {
+		portal.updateName(ctx, *groupChange.ModifyTitle, sender)
+	}
+	if groupChange.ModifyDescription != nil {
+		portal.updateTopic(ctx, *groupChange.ModifyDescription, sender)
+	}
+	if groupChange.ModifyAvatar != nil {
+		portal.updateAvatarWithInfo(ctx, source, groupChange, sender)
+	}
+	intent := sender.IntentFor(portal)
+	for _, addMember := range groupChange.AddMembers {
+		puppet := portal.bridge.GetPuppetBySignalID(addMember.UserID)
+		if puppet == nil {
+			log.Warn().Stringer("signal_user_id", addMember.UserID).Msg("Couldn't get puppet for added member")
+			continue
+		}
+		reqInviteUser := mautrix.ReqInviteUser{
+			Reason: "",
+			UserID: puppet.MXID,
+		}
+		_, err = intent.InviteUser(ctx, portal.MXID, &reqInviteUser)
+		if errors.Is(err, mautrix.MForbidden) {
+			if addMember.JoinFromInviteLink {
+				reqInviteUser.Reason = "Joined From Invite Link"
+			} else {
+				reqInviteUser.Reason = fmt.Sprintf("Added by %s", sender.GetDisplayname())
+			}
+			_, err = portal.MainIntent().InviteUser(ctx, portal.MXID, &reqInviteUser)
+		}
+		if err == nil {
+			err = puppet.IntentFor(portal).EnsureJoined(ctx, portal.MXID)
+		}
+		if err != nil {
+			log.Err(err).Msg("Adding Member failed")
+			return
```

Comment:
Should we actually return here? Could adding a member fail for some reason that would still allow other groupChange events to go through?